frontend/docsettings.lua

```diff
@@ -206,24 +209,17 @@ end
 
 --- Serializes settings and writes them to `metadata.lua`.
 function DocSettings:flush()
-    -- write serialized version of the data table into one of
-    --  i) sidecar directory in the same directory of the document or
-    -- ii) history directory in root directory of KOReader
-    if not self.history_file and not self.sidecar_file then
-        return
+    -- Depending on the settings, doc_settings are saved to the book folder or
+    -- to koreader/sidecars folder. The latter is also a fallback for read-only book storage.
+    local serials = { {self.sidecar_dir_sidecars, self.sidecar_file_sidecars} }
+    if G_reader_settings:readSetting("document_metadata_folder", ":doc") == ":doc" then
+        table.insert(serials, 1, {self.sidecar_dir_doc, self.sidecar_file_doc})
     end
 
-    -- If we can write to sidecar_file, we do not need to write to history_file anymore.
-    local serials = {}
-    if self.sidecar_file then
-        table.insert(serials, self.sidecar_file)
-    end
-    if self.history_file then
-        table.insert(serials, self.history_file)
-    end
-    self:ensureSidecar(self.sidecar)
-    local s_out = dump(self.data)
-    for _, f in ipairs(serials) do
+    local s_out = dump(self.data, nil, true)
+    for _, s in ipairs(serials) do
+        util.makePath(s[1])
+        local f = s[2]
```

Comment:
may be give that `f` a more explicite name :) makes it easier to read below stuff like `self:purge(false, f)`.
`sidecar_file` ?